SudoScript.Core/Solver.cs

```diff
@@ -71,14 +72,96 @@ private static bool SolveRec(Board board, [NotNullWhen(true)] out Board? solvedB
+    {
+        HashSet<Board> result = SolveRecAll(board, 0, limit, random);
+        // If a limit is specified and the result count exceeds it, return only the correct number of solutions.
+        if (limit != 0 && result.Count > limit)
+        {
+            return result.Take(limit).ToList();
+        }
+        return result.ToList();
+    }
+
+    private static HashSet<Board> SolveRecAll(Board board, int solutionCount, int limit, bool random)
+    {
+        HashSet<Board> solutions = new HashSet<Board>();
+        // Eliminate candidates from all rules until nothing changes.
+        while (board.EliminateCandidates());
+        // We hit an invalid state, and must backtrack.
+        if (!board.Validate())
+        {
+            return solutions;
+        }
+        // If the board is solved, return it.
+        if (board.IsSolved())
+        {
+            return new HashSet<Board> { board };
+        }
+
+        // Pick cell to collapse.
+        Cell cell;
+        if (random)
+        {
+            int randomIndex = new Random().Next(0, board.Cells().Count());
+            cell = board.Cells().Skip(randomIndex).First();
+        }
+        else
+        {
+            IEnumerable<Cell> orderedCells = board.Cells()
+                .OrderBy(c => c.CandidateCount);
+            // Skip all cells with less than 2 candidates.
+            orderedCells = orderedCells.SkipWhile(c => c.CandidateCount <= 1);
+            cell = orderedCells.First();
+        }
+
+        foreach (int candidate in cell.Candidates())
+        {
+            // Return if the limit of solutions has been reached
+
+            if ((limit != 0) && ((solutions.Count() + solutionCount) >= limit))
+            {
+                return solutions;
+            }
+
+            // Create a clone of the board for backtracking.
+            Board clonedBoard = board.Clone();
+            Cell clonedCell = clonedBoard[cell.X, cell.Y];
+
+            // Collapse the cell with a digit.
+            clonedCell.Digit = candidate;
+
+            // Call solve on the new board.
+            HashSet<Board> subSolutions = SolveRecAll(clonedBoard, (solutionCount + solutions.Count()), limit, random);
+            solutions.UnionWith(subSolutions);
+        }
+        return solutions;
+    }
+
     public static Board GenerateSolveable(Board board)
     {
         throw new NotImplementedException();
     }
 
-    public static bool IsSatisfactory(Board board)
+    /// <summary>
+    /// Checks if the board can be solved by just using the EliminateCandidates methods from units.
+    /// </summary>
+    /// <param name="board"></param>
+    /// <returns>True if the board can be solved without trial and error guessing.</returns>
+    public static bool IsSatisfactory(Board board) // Certain methods for eliminating candidates using inference are not currently implemented. Implementing them would make this function more acurate.
```

Comment:
Wouldn't it be better to check this by checking all solutions and seeing if there is only a single solution? If there is only a single solution that solution could theoretically always be infered (by simulating infinite backtracking which could be done even on paper).

Reply:
I believe you may be confusing satisfactory with proper

Reply:
Those two terms actually mean the same as i explained in my comment.

If there is only one solution you can always reach said solution without guess work. If you cannot reach the one solution without guess work that means there are more solutions.